### PR TITLE
[Driver] Clean up fp-contract handling in clang driver

### PR DESCRIPTION
**Pull request by andykaylor**
This change refactors the fp-contract handling in
RenderFloatingPointOptions in the clang driver code and fixes some
inconsistencies in the way warnings are reported for changes in the
fp-contract behavior.

### CODE REVIEW

**Comment by andykaylor** [comment]
This is a refreshed version of https://github.com/llvm/llvm-project/pull/91271. Given how long it's been since I updated that PR, it seemed better to just start fresh.

File: clang/test/Driver/fp-contract.c [hunk 1]
```diff
@@ -10,18 +10,18 @@
 // RUN: %clang -### -fno-fast-math -c %s 2>&1 \
 // RUN:   | FileCheck --check-prefix=CHECK-FPC-ON %s
 
-// RUN: %clang -### -Werror -ffast-math -ffp-contract=on -c %s 2>&1 \
+// RUN: %clang -### -ffast-math -ffp-contract=on -c %s 2>&1 \
```
Comment (MaskRay) [inline clang/test/Driver/fp-contract.c:13]: If these RUN lines now emit warnings (upgraded to errors if -Werror), consider adding check prefixes for the warnings.

**Comment by zahiraam** [comment]
> This is a refreshed version of #91271. Given how long it's been since I updated that PR, it seemed better to just start fresh.

May be #91271 should be closed then,
In `Driver/fp-contract.c`, all the `//RUN` lines where you removed the `-Werror` should now have `//WARN` check lines.

File: clang/test/Driver/fp-model.c [hunk 2]
```diff
@@ -81,6 +81,20 @@
...
+// RUN: %clang -### -ffp-model=precise -ffp-contract=fast -c %s 2>&1 \
+// RUN:   | FileCheck --check-prefix=WARN15 %s
+// WARN15: warning: overriding '-ffp-model=precise' option with '-ffp-contract=fast' [-Woverriding-option]
+
+// RUN: %clang -### -ffp-model=strict -fassociative-math -ffp-contract=on \
+// RUN:   -c %s 2>&1 | FileCheck --check-prefix=WARN16 %s
+// WARN16: warning: overriding '-ffp-model=strict' option with '-fassociative-math' [-Woverriding-option]
+// WARN16: warning: overriding '-ffp-model=strict' option with '-ffp-contract=on' [-Woverriding-option]
+
+
```
Comment (MaskRay) [inline clang/test/Driver/fp-model.c:97]: unneeded blank line

**Review by MaskRay — APPROVED** [review]
The logic is complex and difficult to follow, but the test updates looks good!

**Review by zahiraam — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**